### ISSUE TITLE
Custom fixture teardown isn't called

### ISSUE THREAD

**Issue by ben-styling** [vitest-dev/vitest#4251]
### Describe the bug

The docs suggest using text.extend will allow setup and teardown within the fixture.

https://github.com/vitest-dev/vitest/blob/main/docs/guide/test-context.md#usage

I may have set this up incorrectly, but the teardown doesn't seem to fire after the test finishes.

### Reproduction

Stackblitz reproduction, instructions to see error in README.md

https://stackblitz.com/edit/node-3ln3ly?file=.stackblitzrc,example.test.ts,package.json,todos.json,README.md

### System Info

```shell
System:
    OS: macOS 13.2.1
    CPU: (8) arm64 Apple M1 Pro
    Memory: 107.80 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 20.5.0 - ~/.nvm/versions/node/v20.5.0/bin/node
    npm: 10.1.0 - ~/node_modules/.bin/npm
    pnpm: 8.8.0 - ~/Library/pnpm/pnpm
    Watchman: 2023.08.28.00 - /opt/homebrew/bin/watchman
  Browsers:
    Chrome: 117.0.5938.88
    Safari: 16.3
  npmPackages:
    vite: ^4.4.11 => 4.4.11 
    vitest: ^0.34.6 => 0.34.6
```

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by sheremet-va** [comment]
This looks like a bug! @Dunqing 👀

**Comment by ben-styling** [comment]
After a bit of digging, it looks like it might be related to this change #4168 as this works in the last patch.

I would have expected this test to catch this bug! https://github.com/vitest-dev/vitest/blob/ec2e80406555d1eaceefc261d7a50675c2363882/test/core/test/test-extend.test.ts#L303C21-L303C21

**Comment by Dunqing** [comment]
This is a bug and will fixed in #4250. There is a workaround for you can write a test inside `describe` will work
